cloudinit/sources/DataSourceCloudCIX.py

```diff
@@ -0,0 +1,105 @@
+import json
+
+from cloudinit import dmi
+from cloudinit import log as logging
+from cloudinit import sources, url_helper, util
+
+LOG = logging.getLogger(__name__)
+
+DS_BASE_URL = "http://169.254.169.254/"
+
+URL_TIMEOUT = 10
+URL_RETRIES = 5
+URL_SEC_BETWEEN_RETRIES = 1
+
+
+class DataSourceCloudCIX(sources.DataSource):
+
+    dsname = "CloudCIX"
+
+    def __init__(self, sys_cfg, distro, paths):
+        super(DataSourceCloudCIX, self).__init__(sys_cfg, distro, paths)
+
+        self.base_url = url_helper.combine_url(DS_BASE_URL, "v1")
+        self.url_timeout = self.ds_cfg.get("timeout", URL_TIMEOUT)
+        self.url_retries = self.ds_cfg.get("retries", URL_RETRIES)
+        self.wait_retry = self.ds_cfg.get(
+            "wait_retry", URL_SEC_BETWEEN_RETRIES
+        )
+
+    def _get_data(self):
+        if not self.is_running_in_cloudcix():
+            return False
+
+        try:
+            md = util.log_time(
+                LOG.debug,
+                "Crawl of CloudCIX metadata service",
+                read_metadata,
+                kwargs={
+                    "base_url": self.base_url,
+                    "url_params": self.get_url_params(),
+                },
+            )
+        except sources.InvalidMetaDataException as error:
+            LOG.debug(
+                "Failed to read data from CloudCIX datasource: %s", error
+            )
+            return False
+
+        self.metadata = md["meta-data"]
+        self.userdata_raw = md["user-data"]
+        return True
+
+    def is_running_in_cloudcix(self):
```

Comment:
Question for you here: your datasource doesn't define a `DataSourceCloudCIX.network_config property`, so the datasource use the parent class which returns None and therefore your datasource will not write out any network config from your instance metadata. But, you provided `ip_addresses` in your example metadata in unit tests. What configuration do you really want ccloud-init to provide here and will this datasource eventually grow network configuration?

Reply:
Ahh nevermind this question. I hadn't read through your explanation in a prior comment `We're primarily using Cloud-Init for userdata so the metadata is minimal for now`. This works for me for the time being as you grow your instance config needs. One thing to note if you do grow full network config definitions within the metadata, it would be good if you have the ability to render network config v2 format direct from the datasource as cloud-init will be able to provide that configuration unaltered to certain network config renders like `netplan` giving you flexibility in how best to express the network config without cloud-init network primitives transposing/translating a limited set of config options into renderers https://cloudinit.readthedocs.io/en/latest/topics/network-config-format-v2.html#network-config-v2.

Reply:
I'll be able to walk through this again tomorrow to spot check it for unittest coverage, but overall looks pretty good.